### PR TITLE
fix(net/tls) we need to call end when we got FIN if allowHalfOpen is false

### PR DESCRIPTION
**Pull request by cirospaciari**
### What does this PR do?
Fix: https://github.com/oven-sh/bun/issues/13184
Fix: https://github.com/oven-sh/bun/issues/11414

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [X] Code changes

### How did you verify your code works?
Added a test

### CODE REVIEW

**Comment by robobun** [comment]
@cirospaciari, your commit ad221745515a32fe707485936b45756abf321c6a is building: [`#1651`](https://buildkite.com/bun/bun/builds/1651)

File: src/js/node/net.ts [hunk 1]
```diff
@@ -187,17 +187,35 @@ const Socket = (function (InternalSocket) {
...
     static #Close(socket) {
       const self = socket.data;
       if (!self || self.#closed) return;
       self.#closed = true;
       //socket cannot be used after close
       self[bunSocketInternal] = null;
-      const queue = self.#readQueue;
-      if (queue.isEmpty()) {
-        if (self.push(null)) return;
+      if (self.#final_callback) {
```
Comment (Jarred-Sumner) [inline src/js/node/net.ts]:
```suggestion
     const finalCallback = self.#final_callback;
      if (finalCallback) {
        self.#final_callback = null;
        finalCallback()
```

File: test/js/node/net/node-net.test.ts [hunk 2]
```diff
@@ -490,3 +490,36 @@ it("socket should keep process alive if unref is not called", async () => {
   });
   expect(await process.exited).toBe(1);
 });
+
+it.only("should not hang after FIN", async () => {
```
Comment (Jarred-Sumner) [inline test/js/node/net/node-net.test.ts]:
```suggestion
it("should not hang after FIN", async () => {
```

**Comment by wellwelwel** [comment]
Hi @cirospaciari, thanks for this fix 🤝

I have this simple helper:

```ts
import { createConnection } from 'net';

const checkPort = (port: number, host: string): Promise<boolean> =>
  new Promise((resolve) => {
    const client = createConnection(port, host);

    client.on('connect', () => {
      client.destroy(); // it worked normally before this PR
      resolve(true);
    });

    client.on('error', () => {
      resolve(false);
    });
  });
```

---

After this _PR_, the process doesn't finish after a successful port check.

- If I change it to `client.end()` instead of `client.destroy()`, the behavior works as expected.

I also checked the regression changes _(aka `canary`)_, but they don't seem to "visit" the `destroy` method.

---

Do you think it would be interesting to open an issue for this? 🙋🏻‍♂️

**Comment by cirospaciari** [comment]
> ```ts
> import { createConnection } from 'net';
> 
> const checkPort = (port: number, host: string): Promise<boolean> =>
>   new Promise((resolve) => {
>     const client = createConnection(port, host);
> 
>     client.on('connect', () => {
>       client.destroy(); // it worked normally before this PR
>       resolve(true);
>     });
> 
>     client.on('error', () => {
>       resolve(false);
>     });
>   });
> ```

Thanks for reporting and giving the reproduction, I just opened a PR with a fix for that.

**Pull request merged**